File: agents/issue-reconciler/state/leases.ts
import { readFile, writeFile } from 'node:fs/promises';

export type LeaseState = Record<number, string>; // issueNumber -> ISO expiresAt

export function isLeased(state: LeaseState, issueNumber: number, now: Date): boolean {
  const expiresAt = state[issueNumber];
  return expiresAt !== undefined && new Date(expiresAt).getTime() > now.getTime();
}

/** Returns null if already leased and unexpired; otherwise the updated state. */
export function acquireLease(state: LeaseState, issueNumber: number, ttlMs: number, now: Date): LeaseState | null {
  if (isLeased(state, issueNumber, now)) return null;
  return { ...state, [issueNumber]: new Date(now.getTime() + ttlMs).toISOString() };
}

export function releaseLease(state: LeaseState, issueNumber: number): LeaseState {
  const next = { ...state };
  delete next[issueNumber];
  return next;
}

export async function loadLeaseState(path: string): Promise<LeaseState> {
  try {
    return JSON.parse(await readFile(path, 'utf8'));
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') return {};
    throw error;
  }
}

export async function saveLeaseState(path: string, state: LeaseState): Promise<void> {
  await writeFile(path, JSON.stringify(state, null, 2));
}

File: agents/issue-reconciler/issue_reconciler/state/leases.py
from __future__ import annotations

import json
from datetime import datetime, timedelta
from pathlib import Path

# issue_number (as str, since JSON object keys are always strings) -> ISO expires_at
LeaseState = dict[str, str]


def is_leased(state: LeaseState, issue_number: int, now: datetime) -> bool:
    expires_at = state.get(str(issue_number))
    return expires_at is not None and datetime.fromisoformat(expires_at) > now


def acquire_lease(state: LeaseState, issue_number: int, ttl_seconds: int, now: datetime) -> LeaseState | None:
    """Returns None if already leased and unexpired; otherwise the updated state."""
    if is_leased(state, issue_number, now):
        return None
    return {**state, str(issue_number): (now + timedelta(seconds=ttl_seconds)).isoformat()}


def release_lease(state: LeaseState, issue_number: int) -> LeaseState:
    next_state = dict(state)
    next_state.pop(str(issue_number), None)
    return next_state


def load_lease_state(path: Path) -> LeaseState:
    try:
        return json.loads(path.read_text())
    except FileNotFoundError:
        return {}


def save_lease_state(path: Path, state: LeaseState) -> None:
    path.write_text(json.dumps(state, indent=2))
